test/gtest/conv_immediate.cpp

```diff
@@ -0,0 +1,532 @@
+                    desc, pads.size(), pads.data(), strides.data(), dilations.data(), mode) ==
+                miopenStatusSuccess);
+    EXPECT_TRUE(miopenSetConvolutionGroupCount(desc, groupCount) == miopenStatusSuccess);
+
+    return desc;
+}
+
+template <miopen::conv::Direction direction,
+          typename TREF,
+          typename TOUT,
+          miopenTensorLayout_t tensor_layout>
+struct TypeDefs
+{
+    static constexpr miopen::conv::Direction Direction = direction;
+    using TRef                                         = TREF;
+    using TOut                                         = TOUT;
+    static constexpr miopenTensorLayout_t Layout       = tensor_layout;
+};
+
+template <typename T>
+struct layout_data
+{
+    static std::vector<int> get_layout_lengths(int n, int c, std::vector<int>& dims)
+    {
+        auto ret = std::vector<int>{n, c};
+        ret.insert(ret.end(), dims.cbegin(), dims.cend());
+
+        return ret;
+    }
+
+    static std::vector<int>
+    get_strides(std::vector<int>& lens, int dims, miopenTensorLayout_t tensor_layout)
+    {
+        std::vector<int> strides;
+        std::string layout_default = miopen::tensor_layout_get_default(dims + 2);
+        std::string layout_string  = miopen::TensorDescriptor::GetLayoutStr(tensor_layout);
+
+        miopen::tensor_layout_to_strides(lens, layout_default, layout_string, strides);
+
+        constexpr int min_stride_multiplier = 1;
+        constexpr int max_stride_multiplier = 4;
+
+        auto c = prng::gen_A_to_B(min_stride_multiplier, max_stride_multiplier);
+        for(auto& v : strides)
+        {
+            // cppcheck-suppress useStlAlgorithm
+            v = v * c;
+        }
+
+        return strides;
+    }
+
+    static miopenTensorDescriptor_t init_tensor_descriptor(miopenDataType_t type,
+                                                           const std::vector<int>& lens,
+                                                           const std::vector<int>& strides)
+    {
+        miopenTensorDescriptor_t desc;
+
+        EXPECT_TRUE(miopenCreateTensorDescriptor(&desc) == miopenStatusSuccess);
+        EXPECT_TRUE(
+            miopenSetTensorDescriptor(desc, type, lens.size(), lens.data(), strides.data()) ==
+            miopenStatusSuccess);
+
+        return desc;
+    }
+
+    layout_data(int _n, std::vector<int> _dims, int _c, miopenTensorLayout_t _tensor_layout)
+    {
+        auto lens    = get_layout_lengths(_n, _c, _dims);
+        auto strides = get_strides(lens, _dims.size(), _tensor_layout);
+        descriptor   = miopen::TensorDescriptor{miopen_type<T>{}, lens, strides};
+        host         = tensor<T>{lens, strides}.generate(gen_value<T>);
+    }
+
+    ~layout_data() {}
+
+    void read_gpu_data(miopen::Handle& handle, const miopen::Allocator::ManageDataPtr& ddata)
+    {
+        check      = tensor<T>{descriptor.GetLengths(), descriptor.GetStrides()};
+        check.data = handle.Read<T>(ddata, check.data.size());
```

Comment:
Nice, I didn't know about ReadTo. Any time difference wasn't noticeable, but it's still a warm and fuzzy.

Reply:
Sure, it's not a problem for tiny tensors. As well as replacing `std::vector<T>` by `const std::vector<T>&` in the `vet_to_tuple` won't show you any difference.

Since you've provided some abstraction, it would be nice to have a good abstraction without obvious pitfalls, or those pitfalls must be visible and clearly documented.